### ISSUE TITLE
AdEMAMix: Adaptive Exponential Moving Average Mix Optimizer

### ISSUE THREAD

**Issue by ovuruska** [pytorch/pytorch#135609]
### 🚀 The feature, motivation and pitch

AdEMAMix combines Adam and EMA optimization techniques. It addresses slow convergence and poor generalization in large language models and noisy datasets using three beta parameters and an alpha parameter for flexible momentum and adaptive learning rates.

### Alternatives

We considered using existing optimizers with custom learning rate schedules and separate EMA implementations. AdEMAMix unifies adaptive learning rates and exponential moving averages in a single optimizer for better performance and easier implementation.

### Additional context

AdEMAMix showed promising results in large language models, computer vision tasks with noisy labels, and reinforcement learning with sparse rewards. Integration into PyTorch's optimizer suite would benefit researchers working on challenging deep learning problems. 

It introduces new moment to existing AdamW optimizer.

[Implementation](https://github.com/pytorch/pytorch/pull/135610)

cc @vincentqb @jbschlosser @albanD @janeyx99 @crcrpar

**Comment by PhilipMay** [comment]
Hi @ovuruska . IMO AdEMAMix is very interesting and I would like to try it for LLM training. To me (and I guess for many others) it is unclear what you exactly want to say by opening this GitHub issue. Are you requesting anybody to implement it? Do you suggest to contribute / implement it yourself? Perhaps you want to clarify this.

Many thanks, Philip

**Comment by ovuruska** [comment]
Hi @PhilipMay , first of all I share same excitment with this method. I opened this issue to link implementation [here](https://github.com/pytorch/pytorch/pull/135610)

Thanks for asking though. 🙏

**Comment by ovuruska** [comment]
Hi @PhilipMay , first of all I share same excitment with this method. I opened this issue to link implementation [here](https://github.com/pytorch/pytorch/pull/135610)


P.S.: After our conversation, I also added PR link to issue description.

**Comment by janeyx99** [comment]
@ovuruska Thanks for bringing this optimizer to our attention! PyTorch Core has some pretty strict rules when it comes to adding a new Optimizer or Module: https://github.com/pytorch/pytorch/wiki/Developer-FAQ#i-have-a-new-function-or-feature-id-like-to-add-to-pytorch-should-it-be-in-pytorch-core-or-a-library-like-torchvision

At this time, the way to go may be to put it in a separate library and, of course, please don't hesitate to let us know if there is any low level feature that you're missing to implement this externally. Additionally, according to my naive understanding of this optimizer, it would be more generalized/scalable if the EMA portion was more easily applied to all optimizers if possible without needing to add more and more Optimizer classes to our codebase with duplication.

**Comment by ovuruska** [comment]
@janeyx99 Appreciate the feedback. I'll consider your suggestion to reuse EMA. 🫡

**Comment by ZohaibDurrani** [comment]
I have made a different trials on AdEMAMix vs AdamW on same dataset and in same settings. Still the AdamW performed better than AdEMAMix.
